File: server/plan.ts
export type Scalar = "ID" | "String" | "Int" | "Float" | "Boolean";

export interface FieldPlan {
  name: string;
  responseName: string;
  coordinate: string;
  type: string;
  required: boolean;
  list: boolean;
  args: Record<string, { variable: string } | { literal: string | number | boolean | null }>;
  fields?: FieldPlan[];
}

export interface OperationPlan {
  name: string;
  variables: { name: string; type: Scalar; required: boolean }[];
  fields: FieldPlan[];
}

File: internal/plan/plan.go
package plan

import "encoding/json"

type Binding struct {
	Variable string `json:"variable,omitempty"`
	Literal  any    `json:"literal,omitempty"`
}

// A null literal must remain distinguishable from a variable in generated metadata.
func (b Binding) MarshalJSON() ([]byte, error) {
	if b.Variable != "" {
		return json.Marshal(struct {
			Variable string `json:"variable"`
		}{b.Variable})
	}
	return json.Marshal(struct {
		Literal any `json:"literal"`
	}{b.Literal})
}

type Field struct {
	Name         string             `json:"name"`
	ResponseName string             `json:"responseName"`
	Coordinate   string             `json:"coordinate"`
	Type         string             `json:"type"`
	Required     bool               `json:"required"`
	List         bool               `json:"list"`
	Args         map[string]Binding `json:"args"`
	Fields       []Field            `json:"fields,omitempty"`
}

type Variable struct {
	Name     string `json:"name"`
	Type     string `json:"type"`
	Required bool   `json:"required"`
}

type Operation struct {
	Name      string     `json:"name"`
	Variables []Variable `json:"variables"`
	Fields    []Field    `json:"fields"`
}
